### ISSUE TITLE
NaN Gradients When Using mask_where with False Mask on Tensors Containing inf Values

### ISSUE THREAD

**Issue by L-M-Sherlock** [tracel-ai/burn#3727]
**Describe the bug**
When using `mask_where` with a boolean mask set to `false`, the autodiff system still tracks gradients through the masked tensor that contains `inf` values, resulting in `NaN` gradients even though the masked values should not contribute to the computation.

**To Reproduce**
```rust
use burn::backend::Autodiff;
use burn::backend::ndarray::NdArray;
use burn::backend::ndarray::NdArrayDevice;
use burn::tensor::TensorData;
use burn::{module::Param, tensor::Tensor};

fn clone_weights(weights: &Param<Tensor<Autodiff<NdArray>, 1>>) -> Tensor<Autodiff<NdArray>, 1> {
    <Tensor<Autodiff<NdArray>, 1> as Clone>::clone(weights)
}

fn main() {
    type B = Autodiff<NdArray<f32>>;
    let device = NdArrayDevice::Cpu;
    let weights = Param::from_tensor(Tensor::<B, 1>::from_floats([1.0], &device));
    
    // Create a tensor with inf values
    let unused_state = clone_weights(&weights).div(Tensor::<B, 1>::from_floats([0.0], &device));
    
    // Mask the inf values with false (should not be used)
    let mut used_state = clone_weights(&weights);
    used_state = used_state.mask_where(
        Tensor::from_bool(TensorData::from([false]), &device),
        unused_state,
    );
    
    // Compute loss and gradients
    let target = Tensor::<B, 1>::from_floats([1.0], &device);
    let loss = (used_state - target).abs().mean();
    let gradients = loss.backward();
    let weights_grad = weights.grad(&gradients).unwrap();
    
    // weights_grad will be NaN instead of the expected gradient
    println!("weights_grad: {:?}", weights_grad);
}
```

1. Create a tensor with `inf` values by dividing by zero
2. Use `mask_where` with a boolean mask set to `false` to mask out the `inf` values
3. Compute a loss using the masked tensor
4. Compute gradients using `backward()`
5. Extract the gradient for the original parameter
6. Observe that the gradient is `NaN` instead of the expected value

**Expected behavior**
When using `mask_where` with a boolean mask set to `false`, the masked values should not contribute to gradient computation. The gradient should be computed normally based only on the unmasked values, resulting in a valid numerical gradient rather than `NaN`.

**Screenshots**
N/A (terminal output shows the issue)

**Desktop (please complete the following information):**
 - OS: macOS 23.5.0
 - Browser: N/A (Rust application)
 - Version: N/A

**Smartphone (please complete the following information):**
 - Device: N/A
 - OS: N/A
 - Browser: N/A
 - Version: N/A

**Additional context**
- Burn version: 0.17.1
- Backend: Autodiff<NdArray<f32>>
- The issue occurs specifically when:
  1. A tensor contains `inf` values (created by division by zero)
  2. That tensor is masked using `mask_where` with a `false` mask
  3. The masked tensor is used in a computation that requires gradients
  4. The autodiff system still tracks gradients through the masked `inf` values

This suggests that the autodiff system may not be properly handling the masking operation during gradient computation, allowing `inf` values to propagate through the computation graph even when they should be masked out.

**Comment by L-M-Sherlock** [comment]
How did I find this bug:

<img width="1630" height="1656" alt="Image" src="https://github.com/user-attachments/assets/66cf1a19-f9bf-48ac-93d5-d8956bf23c99" />

<img width="1572" height="1656" alt="Image" src="https://github.com/user-attachments/assets/105fb39b-caf0-4702-aa08-78691471de0f" />

The final new_stability doesn't include NaN. But the gradient has NaN.

But FSRS is too complex to debug. So I make the minimal reproduction code above.

**Comment by laggui** [comment]
The issue isn’t a bug in `mask_where` itself. The `NaN`s in the gradients come from the input to `mask_where` already containing `Inf` values. In your minimal reproduction, `weights / 0` produces `Inf`, and backprop through a division by zero produces `NaN`.

You can validate that very simply by explicitly creating an inf input instead of dividing by zero on a tracked tensor:
```rust
// Create a tensor with inf values
// let unused_state = weights.clone().div(TestAutodiffTensor::<1>::from_floats([0.0], &device));
let unused_state = TestAutodiffTensor::<1>::from_floats([f32::INFINITY], &device);
```

and the gradients will be 0.

The issue is most likely similar in your code where there is a division by zero that produces an `Inf` value. Not clear from your screenshots, but adding a small eps where the denominator could be zero avoids the invalid operation and helps stabilize things.

```rust
// Create a tensor with inf values
let unused_state = weights.clone().div(TestAutodiffTensor::<1>::from_floats([0.0], &device) + 1e-6);
```

**Comment by L-M-Sherlock** [comment]
Thanks for the suggestion, it works well!
